File: Kernel Simulator/Misc/Screensaver/Displays/DateAndTime.vb

'    Kernel Simulator  Copyright (C) 2018-2022  Aptivi
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports KS.TimeDate

Namespace Misc.Screensaver.Displays
    Public Module DateAndTimeSettings

        Private _dateAndTime255Colors As Boolean
        Private _dateAndTimeTrueColor As Boolean = True
        Private _dateAndTimeDelay As Integer = 1000
        Private _dateAndTimeMinimumRedColorLevel As Integer = 0
        Private _dateAndTimeMinimumGreenColorLevel As Integer = 0
        Private _dateAndTimeMinimumBlueColorLevel As Integer = 0
        Private _dateAndTimeMinimumColorLevel As Integer = 0
        Private _dateAndTimeMaximumRedColorLevel As Integer = 255
        Private _dateAndTimeMaximumGreenColorLevel As Integer = 255
        Private _dateAndTimeMaximumBlueColorLevel As Integer = 255
        Private _dateAndTimeMaximumColorLevel As Integer = 255

        ''' <summary>
        ''' [DateAndTime] Enable 255 color support. Has a higher priority than 16 color support.
        ''' </summary>
        Public Property DateAndTime255Colors As Boolean
            Get
                Return _dateAndTime255Colors
            End Get
            Set(value As Boolean)
                _dateAndTime255Colors = value
            End Set
        End Property
        ''' <summary>
        ''' [DateAndTime] Enable truecolor support. Has a higher priority than 255 color support.
        ''' </summary>
        Public Property DateAndTimeTrueColor As Boolean
            Get
                Return _dateAndTimeTrueColor
            End Get
            Set(value As Boolean)
                _dateAndTimeTrueColor = value
            End Set
        End Property
        ''' <summary>
        ''' [DateAndTime] How many milliseconds to wait before making the next write?
        ''' </summary>
        Public Property DateAndTimeDelay As Integer
            Get
                Return _dateAndTimeDelay
            End Get
            Set(value As Integer)
                If value <= 0 Then value = 1000
                _dateAndTimeDelay = value
            End Set
        End Property
        ''' <summary>
        ''' [DateAndTime] The minimum red color level (true color)
        ''' </summary>
        Public Property DateAndTimeMinimumRedColorLevel As Integer
            Get
                Return _dateAndTimeMinimumRedColorLevel
            End Get
            Set(value As Integer)
                If value <= 0 Then value = 0
                If value > 255 Then value = 255
                _dateAndTimeMinimumRedColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [DateAndTime] The minimum green color level (true color)
        ''' </summary>
        Public Property DateAndTimeMinimumGreenColorLevel As Integer
            Get
                Return _dateAndTimeMinimumGreenColorLevel
            End Get
            Set(value As Integer)
                If value <= 0 Then value = 0
                If value > 255 Then value = 255
                _dateAndTimeMinimumGreenColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [DateAndTime] The minimum blue color level (true color)
        ''' </summary>
        Public Property DateAndTimeMinimumBlueColorLevel As Integer
            Get
                Return _dateAndTimeMinimumBlueColorLevel
            End Get
            Set(value As Integer)
                If value <= 0 Then value = 0
                If value > 255 Then value = 255
                _dateAndTimeMinimumBlueColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [DateAndTime] The minimum color level (255 colors or 16 colors)
        ''' </summary>
        Public Property DateAndTimeMinimumColorLevel As Integer
            Get
                Return _dateAndTimeMinimumColorLevel
            End Get
            Set(value As Integer)
                Dim FinalMinimumLevel As Integer = If(_dateAndTime255Colors Or _dateAndTimeTrueColor, 255, 15)
                If value <= 0 Then value = 0
                If value > FinalMinimumLevel Then value = FinalMinimumLevel
                _dateAndTimeMinimumColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [DateAndTime] The maximum red color level (true color)
        ''' </summary>
        Public Property DateAndTimeMaximumRedColorLevel As Integer
            Get
                Return _dateAndTimeMaximumRedColorLevel
            End Get
            Set(value As Integer)
                If value <= _dateAndTimeMinimumRedColorLevel Then value = _dateAndTimeMinimumRedColorLevel
                If value > 255 Then value = 255
                _dateAndTimeMaximumRedColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [DateAndTime] The maximum green color level (true color)
        ''' </summary>
        Public Property DateAndTimeMaximumGreenColorLevel As Integer
            Get
                Return _dateAndTimeMaximumGreenColorLevel
            End Get
            Set(value As Integer)
                If value <= _dateAndTimeMinimumGreenColorLevel Then value = _dateAndTimeMinimumGreenColorLevel
                If value > 255 Then value = 255
                _dateAndTimeMaximumGreenColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [DateAndTime] The maximum blue color level (true color)
        ''' </summary>
        Public Property DateAndTimeMaximumBlueColorLevel As Integer
            Get
                Return _dateAndTimeMaximumBlueColorLevel
            End Get
            Set(value As Integer)
                If value <= _dateAndTimeMinimumBlueColorLevel Then value = _dateAndTimeMinimumBlueColorLevel
                If value > 255 Then value = 255
                _dateAndTimeMaximumBlueColorLevel = value
            End Set
        End Property
        ''' <summary>
        ''' [DateAndTime] The maximum color level (255 colors or 16 colors)
        ''' </summary>
        Public Property DateAndTimeMaximumColorLevel As Integer
            Get
                Return _dateAndTimeMaximumColorLevel
            End Get
            Set(value As Integer)
                Dim FinalMaximumLevel As Integer = If(_dateAndTime255Colors Or _dateAndTimeTrueColor, 255, 15)
                If value <= _dateAndTimeMinimumColorLevel Then value = _dateAndTimeMinimumColorLevel
                If value > FinalMaximumLevel Then value = FinalMaximumLevel
                _dateAndTimeMaximumColorLevel = value
            End Set
        End Property

    End Module

    Public Class DateAndTimeDisplay
        Inherits BaseScreensaver
        Implements IScreensaver

        Private RandomDriver As Random

        Public Overrides Property ScreensaverName As String = "DateAndTime" Implements IScreensaver.ScreensaverName

        Public Overrides Property ScreensaverSettings As Dictionary(Of String, Object) Implements IScreensaver.ScreensaverSettings

        Public Overrides Sub ScreensaverPreparation() Implements IScreensaver.ScreensaverPreparation
            'Variable preparations
            RandomDriver = New Random
            Console.BackgroundColor = ConsoleColor.Black
            ConsoleWrapper.Clear()
        End Sub

        Public Overrides Sub ScreensaverLogic() Implements IScreensaver.ScreensaverLogic
            ConsoleWrapper.CursorVisible = False
            ConsoleWrapper.Clear()

            'Write date and time
            SetConsoleColor(ChangeDateAndTimeColor)
            WriteWherePlain(RenderDate, ConsoleWrapper.WindowWidth / 2 - RenderDate.Length / 2, ConsoleWrapper.WindowHeight / 2 - 1)
            WriteWherePlain(RenderTime, ConsoleWrapper.WindowWidth / 2 - RenderTime.Length / 2, ConsoleWrapper.WindowHeight / 2)

            'Delay
            SleepNoBlock(DateAndTimeDelay, ScreensaverDisplayerThread)
        End Sub

        ''' <summary>
        ''' Changes the color of date and time
        ''' </summary>
        Function ChangeDateAndTimeColor() As Color
            Dim ColorInstance As Color
            If DateAndTimeTrueColor Then
                Dim RedColorNum As Integer = RandomDriver.Next(DateAndTimeMinimumRedColorLevel, DateAndTimeMaximumRedColorLevel)
                Dim GreenColorNum As Integer = RandomDriver.Next(DateAndTimeMinimumGreenColorLevel, DateAndTimeMaximumGreenColorLevel)
                Dim BlueColorNum As Integer = RandomDriver.Next(DateAndTimeMinimumBlueColorLevel, DateAndTimeMaximumBlueColorLevel)
                ColorInstance = New Color(RedColorNum, GreenColorNum, BlueColorNum)
            ElseIf DateAndTime255Colors Then
                Dim ColorNum As Integer = RandomDriver.Next(DateAndTimeMinimumColorLevel, DateAndTimeMaximumColorLevel)
                ColorInstance = New Color(ColorNum)
            Else
                ColorInstance = New Color(colors(RandomDriver.Next(DateAndTimeMinimumColorLevel, DateAndTimeMaximumColorLevel)))
            End If
            Return ColorInstance
        End Function

    End Class
End Namespace

File: Kernel Simulator/Misc/Screensaver/Displays/DateAndTime.cs
//
// Kernel Simulator  Copyright (C) 2018-2024  Aptivi
//
// This file is part of Kernel Simulator
//
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY, without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

using Terminaux.Writer.ConsoleWriters;
using KS.Misc.Threading;
using Terminaux.Colors;
using Terminaux.Base;
using KS.TimeDate;
using KS.Misc.Reflection;

namespace KS.Misc.Screensaver.Displays
{
    /// <summary>
    /// Settings for DateAndTime
    /// </summary>
    public static class DateAndTimeSettings
    {
        private static bool _dateAndTime255Colors;
        private static bool _dateAndTimeTrueColor = true;
        private static int _dateAndTimeDelay = 1000;
        private static int _dateAndTimeMinimumRedColorLevel = 0;
        private static int _dateAndTimeMinimumGreenColorLevel = 0;
        private static int _dateAndTimeMinimumBlueColorLevel = 0;
        private static int _dateAndTimeMinimumColorLevel = 0;
        private static int _dateAndTimeMaximumRedColorLevel = 255;
        private static int _dateAndTimeMaximumGreenColorLevel = 255;
        private static int _dateAndTimeMaximumBlueColorLevel = 255;
        private static int _dateAndTimeMaximumColorLevel = 255;

        /// <summary>
        /// [DateAndTime] Enable 255 color support. Has a higher priority than 16 color support.
        /// </summary>
        public static bool DateAndTime255Colors
        {
            get
            {
                return _dateAndTime255Colors;
            }
            set
            {
                _dateAndTime255Colors = value;
            }
        }
        /// <summary>
        /// [DateAndTime] Enable truecolor support. Has a higher priority than 255 color support.
        /// </summary>
        public static bool DateAndTimeTrueColor
        {
            get
            {
                return _dateAndTimeTrueColor;
            }
            set
            {
                _dateAndTimeTrueColor = value;
            }
        }
        /// <summary>
        /// [DateAndTime] How many milliseconds to wait before making the next write?
        /// </summary>
        public static int DateAndTimeDelay
        {
            get
            {
                return _dateAndTimeDelay;
            }
            set
            {
                if (value <= 0)
                    value = 1000;
                _dateAndTimeDelay = value;
            }
        }
        /// <summary>
        /// [DateAndTime] The minimum red color level (true color)
        /// </summary>
        public static int DateAndTimeMinimumRedColorLevel
        {
            get
            {
                return _dateAndTimeMinimumRedColorLevel;
            }
            set
            {
                if (value <= 0)
                    value = 0;
                if (value > 255)
                    value = 255;
                _dateAndTimeMinimumRedColorLevel = value;
            }
        }
        /// <summary>
        /// [DateAndTime] The minimum green color level (true color)
        /// </summary>
        public static int DateAndTimeMinimumGreenColorLevel
        {
            get
            {
                return _dateAndTimeMinimumGreenColorLevel;
            }
            set
            {
                if (value <= 0)
                    value = 0;
                if (value > 255)
                    value = 255;
                _dateAndTimeMinimumGreenColorLevel = value;
            }
        }
        /// <summary>
        /// [DateAndTime] The minimum blue color level (true color)
        /// </summary>
        public static int DateAndTimeMinimumBlueColorLevel
        {
            get
            {
                return _dateAndTimeMinimumBlueColorLevel;
            }
            set
            {
                if (value <= 0)
                    value = 0;
                if (value > 255)
                    value = 255;
                _dateAndTimeMinimumBlueColorLevel = value;
            }
        }
        /// <summary>
        /// [DateAndTime] The minimum color level (255 colors or 16 colors)
        /// </summary>
        public static int DateAndTimeMinimumColorLevel
        {
            get
            {
                return _dateAndTimeMinimumColorLevel;
            }
            set
            {
                int FinalMinimumLevel = _dateAndTime255Colors | _dateAndTimeTrueColor ? 255 : 15;
                if (value <= 0)
                    value = 0;
                if (value > FinalMinimumLevel)
                    value = FinalMinimumLevel;
                _dateAndTimeMinimumColorLevel = value;
            }
        }
        /// <summary>
        /// [DateAndTime] The maximum red color level (true color)
        /// </summary>
        public static int DateAndTimeMaximumRedColorLevel
        {
            get
            {
                return _dateAndTimeMaximumRedColorLevel;
            }
            set
            {
                if (value <= _dateAndTimeMinimumRedColorLevel)
                    value = _dateAndTimeMinimumRedColorLevel;
                if (value > 255)
                    value = 255;
                _dateAndTimeMaximumRedColorLevel = value;
            }
        }
        /// <summary>
        /// [DateAndTime] The maximum green color level (true color)
        /// </summary>
        public static int DateAndTimeMaximumGreenColorLevel
        {
            get
            {
                return _dateAndTimeMaximumGreenColorLevel;
            }
            set
            {
                if (value <= _dateAndTimeMinimumGreenColorLevel)
                    value = _dateAndTimeMinimumGreenColorLevel;
                if (value > 255)
                    value = 255;
                _dateAndTimeMaximumGreenColorLevel = value;
            }
        }
        /// <summary>
        /// [DateAndTime] The maximum blue color level (true color)
        /// </summary>
        public static int DateAndTimeMaximumBlueColorLevel
        {
            get
            {
                return _dateAndTimeMaximumBlueColorLevel;
            }
            set
            {
                if (value <= _dateAndTimeMinimumBlueColorLevel)
                    value = _dateAndTimeMinimumBlueColorLevel;
                if (value > 255)
                    value = 255;
                _dateAndTimeMaximumBlueColorLevel = value;
            }
        }
        /// <summary>
        /// [DateAndTime] The maximum color level (255 colors or 16 colors)
        /// </summary>
        public static int DateAndTimeMaximumColorLevel
        {
            get
            {
                return _dateAndTimeMaximumColorLevel;
            }
            set
            {
                int FinalMaximumLevel = _dateAndTime255Colors | _dateAndTimeTrueColor ? 255 : 15;
                if (value <= _dateAndTimeMinimumColorLevel)
                    value = _dateAndTimeMinimumColorLevel;
                if (value > FinalMaximumLevel)
                    value = FinalMaximumLevel;
                _dateAndTimeMaximumColorLevel = value;
            }
        }
    }

    /// <summary>
    /// Display code for DateAndTime
    /// </summary>
    public class DateAndTimeDisplay : BaseScreensaver, IScreensaver
    {

        private string lastRenderedDate = "";
        private string lastRenderedTime = "";

        /// <inheritdoc/>
        public override string ScreensaverName { get; set; } = "DateAndTime";

        /// <inheritdoc/>
        public override void ScreensaverLogic()
        {
            // Get the color and positions
            Color timeColor = ChangeDateAndTimeColor();
            string renderedDate = TimeDateRenderers.RenderDate();
            string renderedTime = TimeDateRenderers.RenderTime();
            int halfConsoleY = (int)(ConsoleWrapper.WindowHeight / 2d);
            int datePosX = ConsoleWrapper.WindowWidth / 2 - renderedDate.Length / 2;
            int timePosX = ConsoleWrapper.WindowWidth / 2 - renderedTime.Length / 2;

            // Clear old date/time
            int oldDatePosX = ConsoleWrapper.WindowWidth / 2 - lastRenderedDate.Length / 2;
            int oldTimePosX = ConsoleWrapper.WindowWidth / 2 - lastRenderedTime.Length / 2;
            TextWriterWhereColor.WriteWhereColor(new string(' ', lastRenderedDate.Length), oldDatePosX, halfConsoleY, timeColor);
            TextWriterWhereColor.WriteWhereColor(new string(' ', lastRenderedTime.Length), oldTimePosX, halfConsoleY + 1, timeColor);

            // Write date and time
            TextWriterWhereColor.WriteWhereColor(renderedDate, datePosX, halfConsoleY, timeColor);
            TextWriterWhereColor.WriteWhereColor(renderedTime, timePosX, halfConsoleY + 1, timeColor);

            // Delay
            lastRenderedDate = renderedDate;
            lastRenderedTime = renderedTime;
            ThreadManager.SleepNoBlock(DateAndTimeSettings.DateAndTimeDelay, ScreensaverDisplayer.ScreensaverDisplayerThread);
        }

        /// <summary>
        /// Changes the color of date and time
        /// </summary>
        public Color ChangeDateAndTimeColor()
        {
            Color ColorInstance;
            if (DateAndTimeSettings.DateAndTimeTrueColor)
            {
                int RedColorNum = RandomDriver.Random(DateAndTimeSettings.DateAndTimeMinimumRedColorLevel, DateAndTimeSettings.DateAndTimeMaximumRedColorLevel);
                int GreenColorNum = RandomDriver.Random(DateAndTimeSettings.DateAndTimeMinimumGreenColorLevel, DateAndTimeSettings.DateAndTimeMaximumGreenColorLevel);
                int BlueColorNum = RandomDriver.Random(DateAndTimeSettings.DateAndTimeMinimumBlueColorLevel, DateAndTimeSettings.DateAndTimeMaximumBlueColorLevel);
                ColorInstance = new Color(RedColorNum, GreenColorNum, BlueColorNum);
            }
            else
            {
                int ColorNum = RandomDriver.Random(DateAndTimeSettings.DateAndTimeMinimumColorLevel, DateAndTimeSettings.DateAndTimeMaximumColorLevel);
                ColorInstance = new Color(ColorNum);
            }
            return ColorInstance;
        }

    }
}
